### ISSUE TITLE
Methods using `Self` aren't assignable to `Callable[[SubclassOfSelf]]`

### ISSUE THREAD

**Issue by vultix** [microsoft/pyright#10568]
**Describe the bug**
Methods using `Self` aren't assignable to `Callable[[SubclassOfSelf]]`

**Code or Screenshots**
I'd expect this to type-check successfully:
```python
from typing import Callable, Self


class Parent:
    def map(self) -> Self:
        return self


class Child(Parent):
    pass

a: Callable[[Parent], Parent] = Parent.map # type checks successfully
b: Callable[[Child], Child] = Parent.map # type error
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so this isn't a bug.

The expression `Parent.map` partially specializes the method. Its type is initially `(Self) -> Self`, but it becomes `(Parent) -> Parent` after binding. 

In the type alias `Mapper`, the type variable `T` is used in both a covariant and contravariant position, so it is considered invariant. That means `(Parent) -> Parent` is not assignable to `Callable[[Child], Child]`.
